SU2_CFD/include/solvers/CAdjTurbSolver.hpp

```diff
@@ -0,0 +1,193 @@
+/*!
+ * \file CAdjTurbSolver.hpp
+ * \brief Headers of the CAdjTurbSolver class
+ * \author F. Palacios, A. Bueno.
+ * \version 7.0.0 "Blackbird"
+ *
+ * SU2 Project Website: https://su2code.github.io
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2019, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "CSolver.hpp"
+#include "../variables/CAdjTurbVariable.hpp"
+
+/*!
+ * \class CAdjTurbSolver
+ * \brief Main class for defining the adjoint turbulence model solver.
+ * \ingroup Turbulence_Model
+ * \author F. Palacios, A. Bueno.
+ */
+class CAdjTurbSolver : public CSolver {
+private:
+  su2double PsiNu_Inf,  /*!< \brief PsiNu variable at the infinity. */
+  *FlowSolution_i,      /*!< \brief Store the flow solution at point i. */
+  *FlowSolution_j;      /*!< \brief Store the flow solution at point j. */
+
+  su2double Gamma;              /*!< \brief Fluid's Gamma constant (ratio of specific heats). */
+  su2double Gamma_Minus_One;    /*!< \brief Fluids's Gamma - 1.0  . */
+
+  CAdjTurbVariable* nodes = nullptr;  /*!< \brief The highest level in the variable hierarchy this solver can safely use. */
+
+  /*!
+   * \brief Return nodes to allow CSolver::base_nodes to be set.
+   */
+  inline CVariable* GetBaseClassPointerToNodes() override { return nodes; }
```

Comment:
Hey @jayantmukho the specifiers are looking good, a quick comment for cases like these where "all" methods of a class seem final.
What I did with the CVariable family was to declare the class itself final e.g. `class CAdjTurbSolver final : public CSolver` and then the methods override as usual, for two reasons:
- From the optimization standpoint (final gives the compiler license to de-virtualize) the methods of a final class are final by definition.
- From the maintenance perspective if a derived class is added in the future we only need to change one line of code (that was a more important aspect for CVariable with its 100's of methods).

I would reserve final for methods in the middle of the hierarchy as I think that sends the stronger message "this method stops here, no need to navigate further". This being said I know you have been doing your own reading, so if this is advice from some C++ guru with many stripes, forget what I said xD.

Reply:
@pcarruscag you are my C++ guru. That makes sense. 

I am basically done with all the virtual specs. Going to make this change now.

Reply:
Ahah that genuinely made me laugh, thanks!